src/packages/engine-core/src/NAPIEngine.ts

```diff
@@ -0,0 +1,547 @@
+  'linux-arm-openssl-1.1.x',
+  'rhel-openssl-1.0.x',
+  'rhel-openssl-1.1.x',
+  'linux-musl',
+  'linux-nixos',
+  'windows',
+  'freebsd11',
+  'freebsd12',
+  'openbsd',
+  'netbsd',
+  'arm',
+]
+export class NAPIEngine implements Engine {
+  private engine?: QueryEngine
+  private setupPromise?: Promise<void>
+  private connectPromise?: Promise<void>
+  private config: EngineConfig
+  private QueryEngine?: QueryEngineConstructor
+  private logEmitter: EventEmitter
+  libQueryEnginePath?: string
+  platform?: Platform
+  datasourceOverrides: Record<string, string>
+  datamodel: string
+  logQueries: boolean
+  connected: boolean
+  beforeExitListener?: (args?: any) => any
+
+  constructor(config: EngineConfig) {
+    this.datamodel = fs.readFileSync(config.datamodelPath, 'utf-8')
+    this.config = config
+    this.connected = false
+    this.logQueries = config.logQueries ?? false
+    this.logEmitter = new EventEmitter()
+    this.datasourceOverrides = this.config.datasources
+      ? this.convertDatasources(this.config.datasources)
+      : {}
+
+    if (this.logQueries) {
+      process.env.LOG_QUERIES = 'y'
+      this.config.logLevel = 'info'
+    }
+    if (config.enableEngineDebugMode) {
+      Debug.enable('*')
+    }
+    this.setupPromise = this.internalSetup()
+  }
+  private async internalSetup(): Promise<void> {
+    this.platform = await this.getPlatform()
+    this.libQueryEnginePath = await this.getLibQueryEnginePath()
+    return this.loadEngine()
+  }
+  private async getPlatform() {
+    if (this.platform) return this.platform
+    const platform = await getPlatform()
+    if (!knownPlatforms.includes(platform)) {
+      // TODO Update Error
+      throw new PrismaClientInitializationError(
+        `Unknown ${chalk.red(
+          'PRISMA_QUERY_ENGINE_BINARY',
+        )} ${chalk.redBright.bold(
+          this.platform,
+        )}. Possible binaryTargets: ${chalk.greenBright(
+          knownPlatforms.join(', '),
+        )} or a path to the query engine binary.
+You may have to run ${chalk.greenBright(
+          'prisma generate',
+        )} for your changes to take effect.`,
+        this.config.clientVersion!,
+      )
+    }
+    return platform
+  }
+
+  private convertDatasources(
+    datasources: DatasourceOverwrite[],
+  ): Record<string, string> {
+    const obj = Object.create(null)
+    for (const { name, url } of datasources) {
+      obj[name] = url
+    }
+    return obj
+  }
+  private async loadEngine(): Promise<void> {
+    if (!this.engine) {
+      if (!this.QueryEngine) {
+        if (!this.libQueryEnginePath) {
+          this.libQueryEnginePath = await this.getLibQueryEnginePath()
+          debug(`using ${this.libQueryEnginePath}`)
+        }
+        try {
+          this.QueryEngine = require(this.libQueryEnginePath).QueryEngine
+        } catch (e) {
+          // How should we handle this
+          throw new Error(e)
+        }
+      }
+      if (this.QueryEngine) {
```

Comment:
Could be a else?

Reply:
No as the first conditional is loading `this.QueryEngine`